lib/ts/recipe/webauthn/types.ts

```diff
@@ -0,0 +1,625 @@
+
+export type TypeNormalisedInputGetOrigin = (input: { request: BaseRequest; userContext: UserContext }) => string; // should return the app name
+
+export type TypeInput = {
+    emailDelivery?: EmailDeliveryTypeInput<TypeWebauthnEmailDeliveryInput>;
+    relyingPartyId?: TypeInputRelyingPartyId;
+    relyingPartyName?: TypeInputRelyingPartyName;
+    getOrigin?: TypeInputGetOrigin;
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeInputRelyingPartyId =
+    | string
+    | ((input: { request: BaseRequest | undefined; userContext: UserContext }) => string);
+
+export type TypeInputRelyingPartyName = string | (() => string);
+
+export type TypeInputGetOrigin = (input: { request: BaseRequest; userContext: UserContext }) => string;
+
+export type RecipeInterface = {
+    // should have a way to access the user email: passed as a param, through session, or using recoverAccountToken
+    // it should have at least one of those 3 options
+    registerOptions(
+        input: {
+            relyingPartyId: string;
+            relyingPartyName: string;
+            origin: string;
+            requireResidentKey: boolean | undefined; // should default to false in order to allow multiple authenticators to be used; see https://auth0.com/blog/a-look-at-webauthn-resident-credentials/
+            // default to 'required' in order store the private key locally on the device and not on the server
+            residentKey: "required" | "preferred" | "discouraged" | undefined;
+            // default to 'preferred' in order to verify the user (biometrics, pin, etc) based on the device preferences
+            userVerification: "required" | "preferred" | "discouraged" | undefined;
+            // default to 'none' in order to allow any authenticator and not verify attestation
+            attestation: "none" | "indirect" | "direct" | "enterprise" | undefined;
+            // default to 5 seconds
+            timeout: number | undefined;
+            tenantId: string;
+            userContext: UserContext;
+        } & (
+            | {
+                  recoverAccountToken: string;
+              }
+            | {
+                  email: string;
+              }
+            | {
+                  session: SessionContainerInterface;
+              }
+        )
+    ): Promise<
+        | {
+              status: "OK";
+              webauthnGeneratedOptionsId: string;
+              rp: {
+                  id: string;
+                  name: string;
+              };
+              user: {
+                  id: string;
+                  name: string; // user email
+                  displayName: string; //user email
+              };
```

Comment:
It says `The WebAuthn User Name (user.name) is a unique identifier within the RP’s system` - but email is not a unique ID, since many passkeys (belonging to the same user) can have the same email?

Reply:
it's fine to have multiple passkeys. it's not ok to have multiple users with same email.